### PR TITLE
Arc - fix bean generic injection for then required and bean type parameters are type variable, but bean one has no bounds

### LINKED ISSUE #3619: Issue with generic bean injection
When there's a generic bean with an unbounded type parameter T [(example)](https://github.com/eclipse/microprofile-metrics/blob/2.0.1/tck/api/src/main/java/io/astefanutti/metrics/cdi/se/ConcurrentGaugedMethodBean.java#L21) and an injection point that declares a particular type for T [(example)](https://github.com/eclipse/microprofile-metrics/blob/2.0.1/tck/api/src/main/java/io/astefanutti/metrics/cdi/se/ConcurrentGaugedMethodBeanTest.java#L93), ARC blows up with `UnsatisfiedResolutionException`. I assume this setup should be valid, because the CDI [spec](https://docs.jboss.org/cdi/spec/2.0/cdi-spec.html#assignable_parameters) says

> A parameterized bean type is considered assignable to a parameterized required type if they have identical raw type and for each parameter:
> (...)
> the required type parameter is an actual type, the bean type parameter is a type variable and the actual type is assignable to the upper bound, if any, of the type variable, or
> (...)

in our case,
- _required type parameter is an actual type_ - `java.lang.Long`
- _the actual type is assignable to the upper bound of the type variable_ - `Long` is assignable to `java.lang.Object` 

The Metrics TCKs, from which I link examples, work without changes with Weld. To make this test work with Quarkus, I had to add an upper bound for the bean (`ConcurrentGaugedMethodBean<T extends Long>`)

### PR DESCRIPTION
**Pull request by manovotn**
…meters are type variable, but bean one has no bounds.

Fixes #3619

### CODE REVIEW

**Comment by manovotn** [comment]
@jmartisk can you give this a go? It should fix the problem you had in the TCKs...

**Comment by jmartisk** [comment]
This caused ambiguous dependencies in some places, can't compile :/

**Comment by manovotn** [comment]
Yea, can see that. I've only ran it through arc tests before sending this, sorry.
I'll check that.

**Review by jmartisk — APPROVED** [review]
Works for me!

File: independent-projects/arc/processor/src/main/java/io/quarkus/arc/processor/BeanResolver.java [hunk 1]
```diff
@@ -111,7 +111,14 @@ boolean matchesNoBoxing(Type requiredType, Type beanType) {
                     throw new IllegalArgumentException("Invalid argument combination " + requiredType + "; " + beanType);
                 }
                 for (int i = 0; i < requiredTypeArguments.size(); i++) {
-                    if (!parametersMatch(requiredTypeArguments.get(i), beanTypeArguments.get(i))) {
+                    Type requiredTypeArg = requiredTypeArguments.get(i);
+                    Type beanTypeArg = beanTypeArguments.get(i);
+                    if (isActualType(requiredTypeArg) && isActualType(beanTypeArg)
+                            && beanTypeArg.name().equals(DotNames.OBJECT)) {
+                        // required type = FooTyped<Long> and bean type is FooTyped<Object>, that is a match
```
Comment (gsmet) [inline independent-projects/arc/processor/src/main/java/io/quarkus/arc/processor/BeanResolver.java]: We don't support `FooType<Number>` though?

**Review by gsmet — COMMENTED** [review]
Added a question out of curiosity. Looks good but I would prefer Martin having a look.

File: independent-projects/arc/processor/src/main/java/io/quarkus/arc/processor/BeanResolver.java (hunk 1 above)
Comment (manovotn) [reply independent-projects/arc/processor/src/main/java/io/quarkus/arc/processor/BeanResolver.java]: That should already work I think (should better re-check this).
The problem in this PR was mainly raw types which arc automagically converted to from `FooTyped` to `FooTyped<Object>`

Comment (manovotn) [reply independent-projects/arc/processor/src/main/java/io/quarkus/arc/processor/BeanResolver.java]: Wrong...that actually shouldn't work and it doesn't. Neither on master nor on this PR.
Nonetheless, this PR, I'll put that in a separate comment.

**Comment by manovotn** [comment]
So this PR is partially wrong because (according to specification), when you have an injection point (a required type) of `FooTyped<Long>` and then you have beans:
a) `FooTyped<T>`
b) `FooTyped<Object>`

Then a) should be assignable because that is effectively a parameterized type `T extends Object`,  whereas b) should not as then you are comparing simply `Long` and `Object` and those are not equal.
This PR fixes some problems but also allows b) to pass, which is wrong.

**Comment by mkouba** [comment]
Superseded by https://github.com/quarkusio/quarkus/pull/3707

**Pull request closed without merging**